### ISSUE TITLE
Openssl 3.0.0 - 3.0.2 PKCS12_parse password errors on z/OS

### ISSUE THREAD

**Issue by ruijterj** [openssl/openssl#18046]
p12 files that can be parsed without issues in OpenSSL1.0.2.q and below won't parse in Openssl3.0.0, 3.0.1 and 3.0.2.
Errors in Openssl3.0.0 and 3.0.1
' 1800071:PKCS12 routines:PKCS12_parse:mac verify failure:crypto/pkcs12/p12_kiss.c:71:'

Errors in 3.0.2:
'1C800064:Provider routines:ossl_cipher_unpadblock:bad decrypt:providers/implementations/ciphers/ciphercommon_block.c:1
error:11800074:PKCS12 routines:PKCS12_pbe_crypt_ex:pkcs12 cipherfinal error:crypto/pkcs12/p12_decr.c:86:maybe wrong password'
11800072:PKCS12 routines:PKCS12_parse:parse error:crypto/pkcs12/p12_kiss.c:

**Comment by t8m** [comment]
You need legacy provider loaded to be able to load most PKCS12 files.

Unfortunately without having examples of such files available we cannot do much.

**Comment by ruijterj** [comment]
I actually do load those when I initialize the module that calls PKCS12_pars

**Comment by ruijterj** [comment]
"Unfortunately without having examples of such files available we cannot do much." 
Which kind of files do you mean? What do you need?

**Comment by t8m** [comment]
I've meant the PKCS12 files.

Are you able to load these files on some Linux/Windows system? If so, it is most probably something z/OS specific. Perhaps an EBCDIC handling issue?

**Comment by ruijterj** [comment]
I can load them on Linux AND on z/OS using the Openss 1.0.2q code. So indeed my guess would be that something changed in Openssl 3.0.x  that probably is related to ebcdic, otherwise more people would have reported an issue.  We do the conversion from ebcdic to ascii before calling PKCS12_Parse and then things used to work find,but now they no longer do.

**Comment by t8m** [comment]
> We do the conversion from ebcdic to utf8 before calling PKCS12_Parse

What do you convert? The passphrase? Could you try to not convert it?

**Comment by ruijterj** [comment]
yes indeed the passphrase and yes I have tried that, but that resulted in the same error

**Comment by ruijterj** [comment]
the exact point where I hit an error actually appears to be here:

PKCS12_set_mac 

    X509_SIG_getm(p12->mac->dinfo, NULL, &macoct);
    if (!ASN1_OCTET_STRING_set(macoct, mac, maclen)) {
        ERR_raise(ERR_LIB_PKCS12, PKCS12_R_MAC_STRING_SET_ERROR);


Any ideas  ?

**Comment by sjaeckel** [comment]
@t8m this is a failing example https://github.com/strophe/libstrophe/blob/master/tests/cert.nopass.pfx

**Comment by t8m** [comment]
@sjaeckel Where this is failing? This issue is specifically for failures on z/OS.

**Comment by ruijterj** [comment]
@t8m not if it fails on all platforms :-)

**Comment by ruijterj** [comment]
@sjaeckel thank you

**Comment by t8m** [comment]
Please note we are looking for community help implementing this issue. Please comment if you are interested in implementing this feature or if you have reasons why this feature is important enough so we should consider implementing it with higher priority than other requested features. The issue will be closed after 2 months otherwise.

**Comment by t8m** [comment]
This issue was closed because it was marked with the help wanted label and there was no interest in implementing this issue or any other activity for more than 2 months.
